addons/test_mail/tests/test_mail_followers_unsubscribe.py

```diff
@@ -46,3 +53,148 @@ def test_inbox_notification_follower(self):
+            return self._mails[0]['body']
+
+    def _url_tampering(self, url, **kwargs):
+        parsed_url = urlparse(url)
+        return parsed_url._replace(query=urlencode(dict(parse_qsl(parsed_url.query), **kwargs))).geturl()
+
+    def _test_tampered_unfollow_url(self, record, unfollow_url, partner):
+        for param, value in (('token', '0000000000000000000000000000000000000000'),
+                             ('model', 'mail.test.gateway'),
+                             ('res_id', record.copy().id)):
+            with self.subTest(f'Tampered {param}'):
+                tampered_unfollow_url = self._url_tampering(unfollow_url, **{param: value})
+                response = self.url_open(tampered_unfollow_url)
+                self.assertEqual(response.status_code, 403)
+                self.assertIn(partner, record.message_partner_ids)
+
+        with self.subTest('Tampered partner id'):
+            record._message_subscribe(partner_ids=self.partner_admin.ids)
+            tampered_unfollow_url = self._url_tampering(unfollow_url, pid=self.partner_admin.id)
+            response = self.url_open(tampered_unfollow_url)
+            self.assertEqual(response.status_code, 403)
+            self.assertIn(partner, record.message_partner_ids)
+            self.assertIn(self.partner_admin, record.message_partner_ids)
+            record.message_unsubscribe(partner_ids=self.partner_admin.ids)
+
+    def _test_unfollow_url(self, record, unfollow_url, partner, has_record_link=True):
+        with self.subTest('Legitimate unfollow'):
+            # We test that the URL still work a second time if the user has been re-added
+            for _ in range(2):
+                try:
+                    self.assertIn(partner, record.message_partner_ids)
+                    response = self.url_open(unfollow_url)
+                    self.assertEqual(response.status_code, 200)
+                    self.assertNotIn(partner, record.message_partner_ids.ids)
+                    self.assertEqual(urlparse(response.url).path, '/mail/unfollow')
+                    self.assertIn("You are no longer following the document", response.text)
+                    if has_record_link:
+                        self.assertIn('o_access_record_link', response.text)
+                    else:
+                        self.assertNotIn('o_access_record_link', response.text)
+                finally:
+                    record._message_subscribe(partner_ids=partner.ids)
+
+    @users('employee')
+    @mute_logger('odoo.addons.base.models', 'odoo.addons.mail.controllers.mail', 'odoo.http', 'odoo.models')
+    def test_unfollow_internal_user(self):
+        """Internal user must receive an unfollow URL, that works, cannot be tampered and redirects to the correct page.
+
+        Note: When connected, the user must be redirected to the record otherwise a simple front end page.
+        """
+        test_record = self.test_record_admin
```

Comment:
Why are we using record browse as admin in a test using employee as user ? Would prefer making things explicit (test_record_user, and if necessary rebrowse as admin if you want to test a modification done by admin).

Reply:
Yes that’s not logical. I have removed the user annotation because it’s not useful. All the call are done by admin and the checks based on specific user are done through browsing URL either being authenticated or anonymous. I have rewritten the test a bit and added some comments.